lightning-transaction-sync/src/electrum.rs

```diff
@@ -0,0 +1,455 @@
+							let check_tip_hash = notification.header.block_hash();
+							if check_tip_hash != tip_header.block_hash() {
+								tip_header = notification.header;
+								tip_height = notification.height as u32;
+								continue;
+							}
+						}
+
+						num_confirmed += confirmed_txs.len();
+						self.sync_confirmed_transactions(
+							&mut sync_state,
+							&confirmables,
+							confirmed_txs,
+						);
+					}
+					Err(InternalError::Inconsistency) => {
+						// Immediately restart syncing when we encounter any inconsistencies.
+						log_debug!(self.logger,
+							"Encountered inconsistency during transaction sync, restarting.");
+						sync_state.pending_sync = true;
+						continue;
+					}
+					Err(err) => {
+						// (Semi-)permanent failure, retry later.
+						log_error!(self.logger, "Failed during transaction sync, aborting.");
+						sync_state.pending_sync = true;
+						return Err(TxSyncError::from(err));
+					}
+				}
+				sync_state.last_sync_hash = Some(tip_header.block_hash());
+				sync_state.pending_sync = false;
+			}
+		}
+		log_debug!(self.logger,
+			"Finished transaction sync at tip {} in {}ms: {} confirmed, {} unconfirmed.",
+			tip_header.block_hash(), start_time.elapsed().as_millis(), num_confirmed,
+			num_unconfirmed);
+		Ok(())
+	}
+
+	fn sync_unconfirmed_transactions(
+		&self, sync_state: &mut SyncState, confirmables: &Vec<&(dyn Confirm + Sync + Send)>,
+		unconfirmed_txs: Vec<Txid>,
+	) {
+		for txid in unconfirmed_txs {
+			for c in confirmables {
+				c.transaction_unconfirmed(&txid);
+			}
+
+			sync_state.watched_transactions.insert(txid);
+		}
+	}
+
+
+	fn sync_best_block_updated(
+		&self, confirmables: &Vec<&(dyn Confirm + Sync + Send)>, tip_header: &BlockHeader,
+		tip_height: u32,
+	) {
+		for c in confirmables {
+			c.best_block_updated(&tip_header, tip_height);
+		}
+	}
+
+	fn sync_confirmed_transactions(
+		&self, sync_state: &mut SyncState, confirmables: &Vec<&(dyn Confirm + Sync + Send)>,
+		confirmed_txs: Vec<ConfirmedTx>
+	) {
+		for ctx in confirmed_txs {
+			for c in confirmables {
+				c.transactions_confirmed(
+					&ctx.block_header,
+					&[(ctx.pos, &ctx.tx)],
+					ctx.block_height,
+				);
+			}
+
+			sync_state.watched_transactions.remove(&ctx.tx.txid());
+
+			for input in &ctx.tx.input {
+				sync_state.watched_outputs.remove(&input.previous_output);
+			}
+		}
+	}
+
+	fn get_confirmed_transactions(
+		&self, sync_state: &SyncState,
+	) -> Result<Vec<ConfirmedTx>, InternalError> {
+
+		// First, check the confirmation status of registered transactions as well as the
+		// status of dependent transactions of registered outputs.
+		let mut confirmed_txs = Vec::new();
+		let mut watched_script_pubkeys = Vec::new();
+		let mut watched_txs = Vec::new();
+		let watched_outputs = sync_state.watched_outputs.values().collect::<Vec<_>>();
+
+		for txid in &sync_state.watched_transactions {
+			if let Ok(tx) = self.client.transaction_get(&txid) {
```

Comment:
Is the fetch required here because we forget about the transaction once confirmed, but if unconfirmed we only know the txid?